File: 01/main.cxx
#include <iostream>
#include <string>
#include <string_view>
#include <array>
#include <vector>
#include <algorithm>

constexpr int ASCII_0 = 48;
const auto DIGIT_NAMES = std::array<std::string, 10>{
    "zero",
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
};

int parseStringNum(const std::string_view s, int pos)
{
    pos += 1;

    int i = 0;
    for (std::string n : DIGIT_NAMES)
    {
        int begin = std::max(pos - static_cast<int>(n.length()), 0);
        int len = pos - begin;

        auto sub = s.substr(begin, len);

        if (sub == n)
            return i;

        i += 1;
    }

    return -1;
}

int parseDigit(const char c)
{
    if (c >= '0' and c <= '9')
        return c - ASCII_0;

    return -1;
}

std::vector<int> calculateP1(std::string_view s)
{
    std::vector<int> result = {};

    for (const auto c : s)
    {
        auto n = parseDigit(c);
        if (n >= 0)
            result.push_back(n);
    }

    return result;
}

std::vector<int> calculateP2(std::string_view s)
{
    std::vector<int> result = {};

    for (int i = 0; i < s.length(); i += 1)
    {
        auto n = parseDigit(s[i]);
        if (n >= 0)
        {
            result.push_back(n);
            continue;
        }

        n = parseStringNum(s, i);
        if (n >= 0)
            result.push_back(n);
    }

    return result;
}

int solveNumVector(const std::vector<int> &vn)
{
    return 10 * vn.at(0) + vn.back();
}

int main(int argc, char const *argv[])
{
    std::string l{};
    int total_p1_sum = 0;
    int total_p2_sum = 0;

    while (std::getline(std::cin, l))
    {
        total_p1_sum += solveNumVector(calculateP1(l));
        total_p2_sum += solveNumVector(calculateP2(l));
    }

    std::cout << "P1: " << total_p1_sum << '\n';
    std::cout << "P2: " << total_p2_sum << '\n';
    return 0;
}

File: 01/go/main.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"log"
	"os"
	"strings"
)

var numberStrings = []string{
	"zero",
	"one",
	"two",
	"three",
	"four",
	"five",
	"six",
	"seven",
	"eight",
	"nine",
}

var data []string = make([]string, 0, 1000)

func parseInput(r io.Reader) {
	scanner := bufio.NewScanner(r)
	for i := 1; scanner.Scan(); i++ {
		if scanner.Err() != nil {
			log.Fatal(scanner.Err())
		}
		data = append(data, scanner.Text())
	}
}

func extractNumbers(l string, countSpelled bool) (r []int) {
	getDigit := func(c rune) bool {
		if int(c) >= int('0') && int(c) <= int('9') {
			r = append(r, int(c)-int('0'))
			return true
		}

		return false
	}

	getSpelled := func(s string) bool {
		for i, sub := range numberStrings {
			if strings.Contains(s[max(0, len(s)-len(sub)):], sub) {
				r = append(r, i)
				return true
			}

		}

		return false
	}

	for i, c := range l {
		if getDigit(c) {
			continue
		}

		if countSpelled {
			getSpelled(l[max(0, i-4) : i+1])
		}
	}

	return
}

func sumNumbers(countSpelled bool) (r int) {
	decodeDigits := func(d []int) int {
		if len(d) == 0 {
			return 0
		}

		return 10*d[0] + d[len(d)-1]
	}

	for _, l := range data {
		r += decodeDigits(extractNumbers(l, countSpelled))
	}

	return
}
func main() {
	parseInput(os.Stdin)

	P1 := sumNumbers(false)
	P2 := sumNumbers(true)

	fmt.Printf("P1: %d \n", P1)
	fmt.Printf("P2: %d \n", P2)
}
